fastly/service_crud_attribute_definition.go

```diff
@@ -0,0 +1,131 @@
+	// data for the whole service, so the resource argument should be used for most of the creation parameters. The d
+	// argument should just be used for things like getting the service ID. The serviceVersion argument should be used
+	// to decide which version of the service to make updates to. This will be an unlocked version that the base service
+	// update function created.
+	Create(ctx context.Context, d *schema.ResourceData, resource map[string]interface{}, serviceVersion int, conn *gofastly.Client) error
+	// Read should refresh the state of all of the instance of the nested blocks. See the description of Create for more
+	// details about the arguments.
+	Read(ctx context.Context, d *schema.ResourceData, resource map[string]interface{}, serviceVersion int, conn *gofastly.Client) error
+	// Update should make changes to an existing instance of the nested block. The arguments are as described in the
+	// Create comments, with the exception of modified which will contain only the attributes that have changed.
+	Update(ctx context.Context, d *schema.ResourceData, resource, modified map[string]interface{}, serviceVersion int, conn *gofastly.Client) error
+	// Delete should remove the instance of the nested block. See the description of Create for more details about the
+	// arguments.
+	Delete(ctx context.Context, d *schema.ResourceData, resource map[string]interface{}, serviceVersion int, conn *gofastly.Client) error
+}
+
+// ToServiceAttributeDefinition returns an implementation of ServiceAttributeDefinition for a particular implementation
+// of ServiceCRUDAttributeDefinition. It implements the Process and Read methods from ServiceAttributeDefinition using
+// the SetDiff functions.
+func ToServiceAttributeDefinition(definition ServiceCRUDAttributeDefinition) ServiceAttributeDefinition {
+	return &blockSetAttributeHandler{definition}
+}
+
+type blockSetAttributeHandler struct {
+	handler ServiceCRUDAttributeDefinition
+}
+
+func (h *blockSetAttributeHandler) Register(s *schema.Resource) error {
+	s.Schema[h.handler.Key()] = h.handler.GetSchema()
+	return nil
+}
+
+func (h *blockSetAttributeHandler) Read(ctx context.Context, d *schema.ResourceData, s *gofastly.ServiceDetail, conn *gofastly.Client) error {
+	return h.handler.Read(ctx, d, nil, s.ActiveVersion.Number, conn)
+}
+
+func (h *blockSetAttributeHandler) Process(ctx context.Context, d *schema.ResourceData, serviceVersion int, conn *gofastly.Client) error {
+	oldVal, newVal := d.GetChange(h.handler.Key())
+	if oldVal == nil {
+		oldVal = new(schema.Set)
+	}
+	if newVal == nil {
+		newVal = new(schema.Set)
+	}
+
+	oldSet := oldVal.(*schema.Set)
+	newSet := newVal.(*schema.Set)
+
+	setDiff := NewSetDiff(func(resource interface{}) (interface{}, error) {
+		t, ok := resource.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("resource failed to be type asserted: %+v", resource)
+		}
+		return t["name"], nil
```

Comment:
I actually just added this code:
https://github.com/fastly/terraform-provider-fastly/blob/main/fastly/block_fastly_waf_configuration_v1_active_rule.go#L52-L59

which uses a different attribute name than this known "name" attribute. Well, I guess that will be the only place we do such among resources but I wonder if we can make this value dynamic/settable.

Reply:
That's very useful to know. I thought that may be a problem but checked and everything did have `name` attribute so I just left if statically defined as you've spotted. I think it could be settable - by defining a `GetUniqueKeyName` method in the interface or something, and having all the other blocks return "name".

Having said this though, this refactor only affects attributes on the service resource, whereas your change was on the `fastly_waf_configuration` resource and should therefore be isolated. As far as I'm aware, the `ServiceAttributeDefinition` interface is only used in the service resource because of the sheer number of nested blocks. 

If you think it might be likely that either another nested block could be added to the service resource without a `name` attribute, or that we might want to at some point expand the existing `ServiceAttributeDefinition` style of splitting up and composing nested blocks to other resources like `waf_configuration`, then I'm definitely happy to make this change.